### PR TITLE
Dockerfile to create operator registry

### PR DESCRIPTION
**Pull request by baijum**
Task: https://jira.coreos.com/browse/ODC-188

---

### 1. Install OLM (not required for OpenShift 4)

If you are using OpenShift 3, install OLM with this command:

```
kubectl create -f https://raw.githubusercontent.com/operator-framework/operator-lifecycle-manager/master/deploy/upstream/quickstart/olm.yaml 
```

Note: Alternately you can use `oc` command instead of `kubectl.`

### 2. Build and push the operator image to a public registry such as quay.io

Note: Instead of a public registry, the registry provided by OpenShift might work.

Checkout the `master` branch of [devopsconsole-operator](https://github.com/redhat-developer/devopsconsole-operator)

Then run these commands:

```
$ operator-sdk build quay.io/<username>/devopsconsole-operator
$ docker login -u <username> -p <password>  quay.io
$ docker push quay.io/<username>/devopsconsole-operator
```

When running the above command, substitute the `username` and `password` entries appropriately.

### 3. Update the CSV with the operator image location

Open this file
`manifests/devopsconsole/0.1.0/devopsconsole-operator.v0.1.0.clusterserviceversion.yaml` and change the image to point to the location pushed in the previous step.

Inside the file look for `image: REPLACE_IMAGE` and specify the image location.

### 4. Build the operator registry image

Now you are going to build the operator image using `test/olm/Dockerfile.`

```
docker build -f test/olm/Dockerfile . -t quay.io/<username>/operator-registry:0.1.0
docker push quay.io/<username>/operator-registry:0.1.0
```

When running the above command, substitute the `username` with your quay.io username.

### 5. Create CatalogSource and Subscription

Use this template to create a YAML file, say `cat-sub.yaml`:

```
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: my-catalog
  namespace: olm
spec:
  sourceType: grpc
  image: quay.io/<username>/operator-registry:0.1.0
  displayName: Community Operators
  publisher: Red Hat
---
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: my-devopsconsole
  namespace: operators
spec:
  channel: alpha
  name: devopsconsole
  source: my-catalog
  sourceNamespace: olm
```

Before applying the above file, point to the newly created operator registry image (substitute the `username` with your quay.io username).

Example:

```
kubectl apply -f cat-sub.yaml
```

### 6. Verify gitsources CRD presence

Check for the existence of a Custom Resource Definitions with the name as `gitsources.devopsconsole.openshift.io`

Run this command to list CRDs:

```
oc get crds
```

### CODE REVIEW

**Comment by baijum** [comment]
Probably we should keep this Dockerfile at the top directory with a suffix: `Dockerfile.registry` ?

File: test/olm/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+FROM quay.io/openshift/origin-operator-registry:latest
+
+# The operator image location should be updated before copying.
```
Comment (baijum) [inline test/olm/Dockerfile]: I tried to replace pass image location as an argument (`ARG image`) and replace it within the container using `sed`. But I am getting a permission error. If we can fix it, that would be a better approach. Let me know if anyone has got a pointer?

The current approach that I plan to take up is to replace it in the original location and revert it later.

**Comment by corinnekrych** [comment]
@baijum trying to follow the steps as described  in https://jira.coreos.com/browse/ODC-78?focusedCommentId=73032&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-73032
* step1: Install the openshift 4 console ( codenamed 'bridge' ) to run with minishift. from https://github.com/talamer/console which translates into:
```
git clone git@github.com:talamer/console.git
cd console
./build.sh
oc login -u system:admin
oc adm policy add-cluster-role-to-user cluster-admin admin
oc login -u admin
source ./contrib/oc-environment.sh
./bin/bridge
```
open http://localhost:9000/ to see the ui
* step2: Build and push the operator image to quay ( I pushed to quay.io/sbose78/devopsconsole-operator )
which operator image??

**Comment by corinnekrych** [comment]
Thanks @baijum great write up, I'd add it to test/README.md as a step by step installation of the operator in the catalog we can always simplify it once we got it automate in makefile.
It took quite a while before I get a green status in the operator subscription, but all 👍 good.
KEUip5Yw

File: test/olm/Dockerfile (hunk 1 above)
Comment (baijum) [reply test/olm/Dockerfile]: Created an issue to track it for now: #32

**Comment by sbose78** [comment]
@corinnekrych , I was about to comment about having a test/Readme.md :) Will be good to have it linked from the root Readme.md as well.

@baijum this looks good.

**Pull request merged**